### PR TITLE
[#1220] Add front end tests for summary chart functionalities

### PR DESCRIPTION
**Pull request by jamessspanggg**
Fixes part of #1220 

```
Our coverage for front end tests is rather low. UI regressions will be
harder to discover due to this.

Let's add front end tests for summary chart's functionalities such as 
merge group, zoom, filterbreakdown features and the restoring of 
summary controls after page reloads.
```

- merge group
  - [x] when group by repos
  - [x] when group by authors
  - [x] when group by none, will be disabled
- zoom feature
  - [x] clicking on view all commits should open zoom tab
  - [x] zoom into range should open zoom tab
  - [x] zoom into range when merge group should open zoom tab
- filter breakdown feature
  - [x] checking filter breakdown should show file type checkboxes and filetype legends
  - [x] un-checking `all` checkbox will not show any contribution bar and filetype legend
  - [x] un-checking one checkbox will not show that filetype's legend and `all` will be unchecked
- [x] restoring of summary controls after reloading page

### CODE REVIEW

File: frontend/src/summary.pug [hunk 1]
```diff
@@ -42,20 +42,20 @@
...
-          onkeydown="return false", v-bind:min="minDate", v-bind:max="filterUntilDate")
+        input(v-else, type="date", name="since", v-bind:value="filterSinceDate", v-on:input="updateTmpFilterSinceDate",
+          v-bind:min="minDate", v-bind:max="filterUntilDate")
         label since
       .mui-textfield.summary-picker__date
         input(v-if="isSafariBrowser", type="text", placeholder="yyyy-mm-dd",
           v-bind:value="filterUntilDate", v-on:keyup.enter="updateTmpFilterUntilDate",
           onkeydown="formatInputDateOnKeyDown(event)", oninput="appendDashInputDate(event)", maxlength=10)
-        input(v-else, type="date", v-bind:value="filterUntilDate", v-on:input="updateTmpFilterUntilDate",
-          onkeydown="return false", v-bind:min="filterSinceDate", v-bind:max="maxDate")
```
Comment (jamessspanggg) [inline frontend/src/summary.pug:53]: removed this `onkeydown` as there are no other way to test the date picker without typing into it.

**Comment by jamessspanggg** [comment]
@eugenepeh is there a way to suppress the following codacy error? Seems like there's no work around and whenever adding new cypress test cases, codacy will fail.

Screen Shot 2020-04-12 at 6 36 36 PM

File: frontend/cypress/tests/chartView_filterBreakdown.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        .first()
+        .uncheck()
+        .should('not.be.checked');
+
+    // should not show file type legends
+    cy.get('#summary-charts div.summary-charts__fileType--breakdown')
+        .should('not.be.visible');
+
+    // should not show any contribution bar
+    cy.get('#summary-charts div.summary-chart__contribution')
```
Comment (0blivious) [inline frontend/cypress/tests/chartView_filterBreakdown.js:38]: This is overlapped with the test in `contributionBar.js`

**Comment by 0blivious** [comment]
Currently there are cypress tests named `codeView__$` but in your new tests the naming is changed to chartView. I feel it is better to standardise the names. Also if we are having so many tests related to chartView, maybe we should have a folder for it?

**Comment by jamessspanggg** [comment]
> Currently there are cypress tests named codeView__$ but in your new tests the naming is changed to chartView.

chartView refers to the summary chart tab while codeView refers to the authorship tab. But yeah can do, will create folders and place accordingly.

File: frontend/cypress/tests/chartView/chartView_filterBreakdown.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        .uncheck()
+        .should('not.be.checked');
+
+    // should not show file type legends
+    cy.get('#summary-charts div.summary-charts__fileType--breakdown')
+        .should('not.be.visible');
+  });
+
+  it('uncheck file type should uncheck all option and not show legend', () => {
+    Cypress.wait();
```
Comment (Tejas2805) [inline frontend/cypress/tests/chartView/chartView_filterBreakdown.js:39]: Can you add a test for - if one of the checkboxes is not checked, the `All` checkboxes should also not be checked?

Comment (jamessspanggg) [reply frontend/cypress/tests/chartView/chartView_filterBreakdown.js:39]: done

File: frontend/cypress/tests/chartView/chartView_filterBreakdown.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    cy.get('#summary label.filter-breakdown input:visible')
+        .should('be.visible')
+        .check()
+        .should('be.checked');
+
+    // uncheck all gradle file type
```
Comment (Tejas2805) [inline frontend/cypress/tests/chartView/chartView_filterBreakdown.js:46]:
```suggestion
    // uncheck gradle file type
```

File: frontend/cypress/tests/chartView/chartView_zoomFeature.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        .should('be.checked');
+
+    // clicking from the 10th px to the 50th px in the ramp
+    cy.get('body').type(zoomKey, { release: false })
+        .get('#summary-charts .summary-chart__ramp .ramp')
+        .first()
+        .click(10, 20)
+        .click(50, 20);
+
+    cy.get('#tab-zoom')
```
Comment (ccyccyccy) [inline frontend/cypress/tests/chartView/chartView_zoomFeature.js:44]: Maybe can test for reload here as well to see if the zoom selection is preserved?

File: frontend/cypress/tests/chartView/chartView_filterBreakdown.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    Cypress.wait();
+
+    cy.get('#summary label.filter-breakdown input:visible')
+        .should('be.visible')
+        .check()
+        .should('be.checked');
+
+    // uncheck all file types
+    cy.get('#summary div.fileTypes input:visible')
+        .first()
```
[the remark this answers is not in the document]
Comment (jamessspanggg) [reply frontend/cypress/tests/chartView/chartView_filterBreakdown.js]: yes you are right, have done so

File: frontend/cypress/tests/chartView/chartView_zoomFeature.js (hunk 5 above)
Comment (jamessspanggg) [reply frontend/cypress/tests/chartView/chartView_zoomFeature.js:44]: I actually intend to leave that as a separate PR that tests all things related to zoom. Currently, this PR only tests all things related to summary

Comment (jamessspanggg) [reply frontend/cypress/tests/chartView/chartView_zoomFeature.js:44]: or do you mean the dashed lines when activating zoom? those lines are not preserved after reloading currently.

Comment (ccyccyccy) [reply frontend/cypress/tests/chartView/chartView_zoomFeature.js:44]: I was talking about the zoom tab on the right. Wasn't sure if that was considered part of this PR. Guess not.

**Comment by jamessspanggg** [comment]
@eugenepeh @fzdy1914 can I get reviews from you guys?

**Pull request merged**